### PR TITLE
fix: prevent putDimensions from storing duplicate dimensions

### PR DESCRIPTION
**Pull request by aaronlna**
## Issue

https://github.com/awslabs/aws-embedded-metrics-java/issues/35 (related)

## Description of changes

This is the Java-equivalent bug-fix for prior issue in Node:
* https://github.com/awslabs/aws-embedded-metrics-node/issues/20
* https://github.com/awslabs/aws-embedded-metrics-node/pull/104

This change introduces the following behaviors:
1. dimension sets added through `#putDimensions` do not duplicate
2. duplicate dimension sets replace (and are sorted after) existing dimension sets such
that newer/latest dimension values are used by the root EMF node

Example:

```
putDimensions({ "keyA": "valueA1" })
putDimensions({ "keyA": "valueA2" })
putDimensions({ "keyA": "valueA3", "keyB": "valueB1" })
putDimensions({ "keyB": "valueB2", "keyA": "valueA4" })

// expected MetricsDirective dimensions:
[
  { "keyA": "valueA2" },
  { "keyB": "valueB2", "keyA": "valueA4" },
]

// expected EMF target members:
{
  "keyA": "valueA4",
  "keyB": "valueB2",
}
```

## Description of how you validated changes

Unit test updated to address edge case;
multiple dimension sets of variable size are added and checked.

Ran integration tests using Docker and compared results in CloudWatch.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by miguelcss** [comment]
Ping, anyone at aws-embedded-metrics-java project with admin permissions can have a look at this PR? 
Some more coming up :)

File: src/test/java/software/amazon/cloudwatchlogs/emf/model/MetricDirectiveTest.java [hunk 1]
```diff
@@ -114,6 +114,57 @@ public void testPutMultipleDimensionSets() throws JsonProcessingException {
...
+                DimensionSet.of("Region", "us-east-1", "Instance", "inst-1"));
+        metricDirective.putDimensionSet(
+                DimensionSet.of("Instance", "inst-1", "Region", "us-east-1"));
+        metricDirective.putDimensionSet(DimensionSet.of("Instance", "inst-1"));
+
+        String serializedMetricDirective = objectMapper.writeValueAsString(metricDirective);
+
+        assertEquals(
+                serializedMetricDirective,
+                "{\"Dimensions\":[[],[\"Region\"],[\"Instance\",\"Region\"],[\"Instance\"]],\"Metrics\":[],\"Namespace\":\"aws-embedded-metrics\"}");
```
Comment (markkuhn) [inline src/test/java/software/amazon/cloudwatchlogs/emf/model/MetricDirectiveTest.java:139]: Swap these 2 arguments so they are in the correct order: expected value, actual value.

```suggestion
                "{\"Dimensions\":[[],[\"Region\"],[\"Instance\",\"Region\"],[\"Instance\"]],\"Metrics\":[],\"Namespace\":\"aws-embedded-metrics\"}",
                serializedMetricDirective);
```

File: src/test/java/software/amazon/cloudwatchlogs/emf/model/MetricDirectiveTest.java [hunk 2]
```diff
@@ -114,6 +114,57 @@ public void testPutMultipleDimensionSets() throws JsonProcessingException {
...
+                DimensionSet.of("Instance", "inst-1", "Region", "us-east-1"));
+        metricDirective.putDimensionSet(DimensionSet.of("Instance", "inst-1"));
+        metricDirective.putDimensionSet(DimensionSet.of("Region", "us-east-1"));
+        metricDirective.putDimensionSet(new DimensionSet());
+
+        String serializedMetricDirective = objectMapper.writeValueAsString(metricDirective);
+
+        assertEquals(
+                serializedMetricDirective,
+                "{\"Dimensions\":[[\"Instance\",\"Region\"],[\"Instance\"],[\"Region\"],[]],\"Metrics\":[],\"Namespace\":\"aws-embedded-metrics\"}");
```
Comment (markkuhn) [inline src/test/java/software/amazon/cloudwatchlogs/emf/model/MetricDirectiveTest.java:165]: Swap these 2 arguments so they are in the correct order: expected value, actual value.

```suggestion
                "{\"Dimensions\":[[\"Instance\",\"Region\"],[\"Instance\"],[\"Region\"],[]],\"Metrics\":[],\"Namespace\":\"aws-embedded-metrics\"}",
                serializedMetricDirective);
```

**Review by markkuhn — CHANGES_REQUESTED** [review]

**Review by Stephen-Bao — APPROVED** [review]

**Review by markkuhn — APPROVED** [review]
